File: backend/app/schemas/stocks.py
"""Pydantic schemas for the stocks / screener API."""
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel


class StockRow(BaseModel):
    """
    One row of the Screener table. Every numeric field is decimal / annual
    so the frontend renders them consistently. `null` means the indicator
    has not been computed yet — either a fresh stock or a yfinance coverage
    gap. The Screener renders NULL as "—" per the accountant coloring rule.
    """
    # Identity
    symbol: str
    ticker_suffix: str
    name_ar: str | None
    name_en: str | None
    industry_ar: str | None
    industry_en: str | None
    sector_code: str | None

    # Risk indicators (group 1 of 2 in the Screener filter UI)
    beta: float | None
    capm_expected_return: float | None
    daily_volatility: float | None
    annual_volatility: float | None
    sharp_ratio: float | None
    var_95_daily: float | None
    risk_ranking: str | None

    # Financial indicators (group 2)
    pe_ratio: float | None
    market_to_book: float | None
    roe: float | None
    fcf_yield: float | None
    leverage_ratio: float | None
    eps: float | None
    dividend_yield: float | None
    annual_dividend_rate: float | None

    # Price snapshot
    last_price: float | None
    last_price_date: date | None
    last_analytics_refresh: datetime | None

    # Fundamental disclosure dates (rendered in the Screener's Financial
    # Ratios band). None → "—" on the frontend.
    last_balance_sheet_date: date | None
    last_income_statement_date: date | None
    latest_dividend_date: date | None

    model_config = {"from_attributes": True}


class DataSourceRange(BaseModel):
    """Date range + provider info for one data-sources footer card."""
    id: str
    date_from: date | None
    date_to: date | None
    source_name: str
    source_url: str | None


class DataSourcesOut(BaseModel):
    """Payload for the Screener's "مصادر البيانات وفترات التحديث" footer."""
    stock_prices: DataSourceRange
    sector_indices: DataSourceRange
    last_update: DataSourceRange


# ── Stock Analyze page (Loay slides 98-99 / 109-111) ────────────────────────
class ReturnDistributionBucket(BaseModel):
    """One histogram bucket on the stock-return probability chart."""
    lower: float
    upper: float
    frequency_pct: float


class PricePoint(BaseModel):
    trade_date: date
    close: float


class ReturnPair(BaseModel):
    trade_date: date
    stock_return: float
    index_return: float | None


class StockAnalyticsOut(BaseModel):
    """Full payload for the per-stock Analyze page."""
    # Identity
    symbol: str
    ticker_suffix: str
    name_ar: str | None
    name_en: str | None
    sector_code: str | None
    industry_ar: str | None
    industry_en: str | None

    # Stock Movement block
    last_price: float | None
    last_price_date: date | None
    week52_high: float | None
    week52_low: float | None
    avg_price_midpoint: float | None
    min_return_250d: float | None
    max_return_250d: float | None

    # Stock Risk Measurement block
    beta: float | None
    capm_expected_return: float | None
    daily_volatility: float | None
    annual_volatility: float | None
    sharp_ratio: float | None
    var_95_daily: float | None
    risk_ranking: str | None

    # Financial Ratios block — the 14 + the 16 extras from Phase B
    pe_ratio: float | None
    market_to_book: float | None
    roe: float | None
    fcf_yield: float | None
    leverage_ratio: float | None
    eps: float | None
    dividend_yield: float | None
    annual_dividend_rate: float | None
    current_ratio: float | None
    quick_ratio: float | None
    cash_ratio: float | None
    interest_coverage_ratio: float | None
    asset_turnover: float | None
    inventory_turnover: float | None
    receivables_turnover: float | None
    payables_turnover: float | None
    roa: float | None
    net_profit_margin: float | None
    gross_profit_margin: float | None
    book_value_per_share: float | None
    revenue_per_share: float | None
    debt_to_market_cap: float | None
    cash_to_assets: float | None
    receivables_to_assets: float | None

    # Disclosure dates
    last_balance_sheet_date: date | None
    last_income_statement_date: date | None
    latest_dividend_date: date | None

    # Support & resistance (from last 30 trading days)
    support_price: float | None
    resistance_price: float | None
    midpoint_price: float | None

    # Charts: probability distribution + price history + stock-vs-index pair
    return_distribution: list[ReturnDistributionBucket]
    price_history: list[PricePoint]
    stock_vs_index: list[ReturnPair]

    # CAPM-derived expected returns (used by the Expected Stock Return block)
    expected_annual_return: float | None
    expected_daily_return: float | None


class SectorAveragesOut(BaseModel):
    """Sector-level averages of the 14 indicators (Loay's slide 83)."""
    sector_code: str
    sector_name_ar: str
    sector_name_en: str
    stock_count: int
    # Risk
    avg_beta: float | None
    avg_capm_expected_return: float | None
    avg_daily_volatility: float | None
    avg_annual_volatility: float | None
    avg_sharp_ratio: float | None
    avg_var_95_daily: float | None
    risk_ranking: str | None
    # Financial
    avg_pe_ratio: float | None
    avg_market_to_book: float | None
    avg_roe: float | None
    avg_fcf_yield: float | None
    avg_leverage_ratio: float | None
    avg_eps: float | None
    avg_dividend_yield: float | None
    avg_annual_dividend_rate: float | None


class SectorSummary(BaseModel):
    """Lightweight: code + name + stock count, populates the sector picker."""
    sector_code: str
    sector_name_ar: str
    sector_name_en: str
    stock_count: int

File: frontend/src/api/stocks.ts
import { api } from "./client";

/**
 * Full screener row — all 14 indicators from PPTX slide 83.
 * Every numeric is `number | null`; the Screener renders `null` as "—"
 * via `numToneClass(null) → text-muted`.
 */
export interface StockRow {
  // Identity
  symbol: string;
  ticker_suffix: string;
  name_ar: string | null;
  name_en: string | null;
  industry_ar: string | null;
  industry_en: string | null;
  sector_code: string | null;

  // Risk indicators (6)
  beta: number | null;
  capm_expected_return: number | null;
  daily_volatility: number | null;
  annual_volatility: number | null;
  sharp_ratio: number | null;
  var_95_daily: number | null;
  risk_ranking: string | null;

  // Financial indicators (8)
  pe_ratio: number | null;
  market_to_book: number | null;
  roe: number | null;
  fcf_yield: number | null;
  leverage_ratio: number | null;
  eps: number | null;
  dividend_yield: number | null;
  annual_dividend_rate: number | null;

  // Price snapshot
  last_price: number | null;
  last_price_date: string | null;
  last_analytics_refresh: string | null;

  // Disclosure dates (Loay slide — Financial Ratios band)
  last_balance_sheet_date: string | null;
  last_income_statement_date: string | null;
  latest_dividend_date: string | null;
}

// Data sources & update periods footer (Loay slide — مصادر البيانات وفترات التحديث)
export interface DataSourceRange {
  id: string;
  date_from: string | null;
  date_to: string | null;
  source_name: string;
  source_url: string | null;
}

export interface DataSourcesPayload {
  stock_prices: DataSourceRange;
  sector_indices: DataSourceRange;
  last_update: DataSourceRange;
}

// Sector-level summary + averages (Loay slide 83)
export interface SectorSummary {
  sector_code: string;
  sector_name_ar: string;
  sector_name_en: string;
  stock_count: number;
}

export interface SectorAverages {
  sector_code: string;
  sector_name_ar: string;
  sector_name_en: string;
  stock_count: number;
  // Risk
  avg_beta: number | null;
  avg_capm_expected_return: number | null;
  avg_daily_volatility: number | null;
  avg_annual_volatility: number | null;
  avg_sharp_ratio: number | null;
  avg_var_95_daily: number | null;
  risk_ranking: string | null;
  // Financial
  avg_pe_ratio: number | null;
  avg_market_to_book: number | null;
  avg_roe: number | null;
  avg_fcf_yield: number | null;
  avg_leverage_ratio: number | null;
  avg_eps: number | null;
  avg_dividend_yield: number | null;
  avg_annual_dividend_rate: number | null;
}

// Stock Analyze page (Loay slides 98-99 / 109-111)
export interface ReturnDistributionBucket {
  lower: number;
  upper: number;
  frequency_pct: number;
}

export interface PricePoint {
  trade_date: string;
  close: number;
}

export interface ReturnPair {
  trade_date: string;
  stock_return: number;
  index_return: number | null;
}

export interface StockAnalytics {
  // Identity
  symbol: string;
  ticker_suffix: string;
  name_ar: string | null;
  name_en: string | null;
  sector_code: string | null;
  industry_ar: string | null;
  industry_en: string | null;
  // Movement
  last_price: number | null;
  last_price_date: string | null;
  week52_high: number | null;
  week52_low: number | null;
  avg_price_midpoint: number | null;
  min_return_250d: number | null;
  max_return_250d: number | null;
  // Risk
  beta: number | null;
  capm_expected_return: number | null;
  daily_volatility: number | null;
  annual_volatility: number | null;
  sharp_ratio: number | null;
  var_95_daily: number | null;
  risk_ranking: string | null;
  // Financial Ratios — 14 + 16
  pe_ratio: number | null;
  market_to_book: number | null;
  roe: number | null;
  fcf_yield: number | null;
  leverage_ratio: number | null;
  eps: number | null;
  dividend_yield: number | null;
  annual_dividend_rate: number | null;
  current_ratio: number | null;
  quick_ratio: number | null;
  cash_ratio: number | null;
  interest_coverage_ratio: number | null;
  asset_turnover: number | null;
  inventory_turnover: number | null;
  receivables_turnover: number | null;
  payables_turnover: number | null;
  roa: number | null;
  net_profit_margin: number | null;
  gross_profit_margin: number | null;
  book_value_per_share: number | null;
  revenue_per_share: number | null;
  debt_to_market_cap: number | null;
  cash_to_assets: number | null;
  receivables_to_assets: number | null;
  // Disclosure
  last_balance_sheet_date: string | null;
  last_income_statement_date: string | null;
  latest_dividend_date: string | null;
  // Support/Resistance + chart series
  support_price: number | null;
  resistance_price: number | null;
  midpoint_price: number | null;
  return_distribution: ReturnDistributionBucket[];
  price_history: PricePoint[];
  stock_vs_index: ReturnPair[];
  expected_annual_return: number | null;
  expected_daily_return: number | null;
}

export const StocksAPI = {
  list: () => api<StockRow[]>("/stocks"),
  sectorsSummary: () => api<SectorSummary[]>("/stocks/sectors-summary"),
  sectorAverages: (sectorCode: string) =>
    api<SectorAverages>(`/stocks/sector-averages?sector_code=${encodeURIComponent(sectorCode)}`),
  dataSources: () => api<DataSourcesPayload>("/stocks/data-sources"),
  analytics: (ticker: string) =>
    api<StockAnalytics>(`/stocks/${encodeURIComponent(ticker)}/analytics`),
};
